### PR TITLE
[Merged by Bors] - Implement and require `#[derive(Component)]` on all component structs

### PR DESCRIPTION
**Pull request by Frizi**
This implements the most minimal variant of #1843 - a derive for marker trait. This is a prerequisite to more complicated features like statically defined storage type or opt-out component reflection.

In order to make component struct's purpose explicit and avoid misuse, it must be annotated with `#[derive(Component)]` (manual impl is discouraged for compatibility). Right now this is just a marker trait, but in the future it might be expanded. Making this change early allows us to make further changes later without breaking backward compatibility for derive macro users.

This already prevents a lot of issues, like using bundles in `insert` calls. Primitive types are no longer valid components as well. This can be easily worked around by adding newtype wrappers and deriving `Component` for them.

One funny example of prevented bad code (from our own tests) is when an newtype struct or enum variant is used. Previously, it was possible to write `insert(Newtype)` instead of `insert(Newtype(value))`. That code compiled, because function pointers (in this case newtype struct constructor) implement `Send + Sync + 'static`, so we allowed them to be used as components. This is no longer the case and such invalid code will trigger a compile error.

### CODE REVIEW

**Comment by NathanSWard** [comment]
#1843 is tagged with `needs-rfc`. Are we planning on not going down the rfc route for this?

File: crates/bevy_app/src/app_builder.rs
Comment (alice-i-cecile) [inline crates/bevy_app/src/app_builder.rs:255]: I like this change a lot. Dramatically clearer and safer.

File: crates/bevy_app/src/app_builder.rs
Comment (alice-i-cecile) [inline crates/bevy_app/src/app_builder.rs:295]: Yay separating out these bounds! I'll need to swap this to `T: Component + Resource` for #1626, but that shouldn't be any issue.

File: crates/bevy_asset/src/loader.rs
Comment (alice-i-cecile) [inline crates/bevy_asset/src/loader.rs:146]: This surprises me a bit: why were we able to remove this component bound / why did it exist in the first place?

File: crates/bevy_core/src/label.rs
Comment (alice-i-cecile) [inline crates/bevy_core/src/label.rs:18]: Is there any chance we can wrap this into the derive, probably in a follow-up PR? That would make working with reflection much less frustrating.

**Comment by Frizi** [comment]
> #1843 is tagged with `needs-rfc`. Are we planning on not going down the rfc route for this?

I believe there was a green light for this minimal implementation. Though it might have ben said only on discord. I'm explicilty avoiding adding any more features than necessary here, as those might indeed need further discussion and possible RFC.

File: crates/bevy_ecs/src/component/mod.rs
Comment (NathanSWard) [inline crates/bevy_ecs/src/component/mod.rs:34]: Can we create an issue for this (if/when) this PR gets merged?

File: crates/bevy_core/src/label.rs
Comment (Frizi) [reply crates/bevy_core/src/label.rs:18]: Yes, I believe this is one of the first things we might add. But I intentionally kept it as is to limit the scope of changes here. Details around automatic component reflection might need further discussion.

File: crates/bevy_asset/src/loader.rs
Comment (Frizi) [reply crates/bevy_asset/src/loader.rs:146]: I'm not sure exactly. I was surprised about that bound as well, as it didn't make much sense to me. I think it bound should be more like `T: Asset`, but clearly that bound wasn't used either. Removing it seems to be just fine, so I went with that.

File: crates/bevy_ecs/src/component/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/component/mod.rs:17]: I would like a comment here describing how to work around the orphan rules for new Rust users.

File: crates/bevy_ecs/src/component/mod.rs
Comment (Frizi) [reply crates/bevy_ecs/src/component/mod.rs:34]: Sure, it's a good follow-up once those changes are accepted. Unless we decide to get rid of that in this PR anyway :)

**Comment by alice-i-cecile** [comment]
Another follow-up issue that should be created for discussion: require `#[derive(Resource)`.

**Comment by cart** [comment]
> I believe there was a green light for this minimal implementation. Though it might have ben said only on discord. I'm explicilty avoiding adding any more features than necessary here, as those might indeed need further discussion and possible RFC. This PR doesn't close that linked issue.

Hmm I'm not remembering that (but my memory isn't perfect :smile:).  I _am_ pretty sold on this, but my stance is still this: https://github.com/bevyengine/bevy/issues/1843#issuecomment-844358952.

Before merging I'd like an RFC with a clear plan for the future (I'm most interested in what the Component trait will look like and the derive macro interface), and _some_ killer feature implemented to make this more palatable for users. 

As it stands from a user perspective, this pr currently just (1) adds boilerplate and (2) solves a problem that, while common, probably isn't on most people's radar.

I think implementing static storage should be that killer feature, because it:

1. Ideally makes normal for loop Query iteration faster (people respond very favorably to benchmark results). I think showing "optimization graphs" in our release notes is a big part of bevy's positive reception. I think a graph is the "spoon full of sugar" that will make the boilerplate "medicine" go down.
2. Forces us to prove that one of our (or at least, my) biggest reasons for merging (better for loop perf) is grounded in reality 
3. Hopefully makes storage configuration simpler
4. Might optimize other things because we no longer need to store intermediate storage info or do runtime storage type checks.

**Comment by Frizi** [comment]
> Before merging I'd like an RFC with a clear plan for the future (I'm most interested in what the Component trait will look like and the derive macro interface), and some killer feature implemented to make this more palatable for users.

Okay, I'm fine with that. The reason I wanted the minimal version in the first place is due to current efforts to bring relations into ECS. Implementing type-level storages will require deeper changes that would severely conflict with that branch. So, we have a choice to make here.

I can ignore that conflict and start the implementation anyway. This would allow us to measure performance gains quickly, but later will add nontrivial merging overhead. Alternatively, this can just wait until relations are merged in.

**Comment by alice-i-cecile** [comment]
I would prefer to do this before relations; it may open up new architectural options for us to help resolve some of the outstanding concerns.

**Comment by cart** [comment]
> I can ignore that conflict and start the implementation anyway. This would allow us to measure performance gains quickly, but later will add nontrivial merging overhead. Alternatively, this can just wait until relations are merged in.

This is a tough call. I think this will be less controversial and easier to review/merge than relations. But no matter what we do somebody is gonna need to deal with a->b or b->a merge conflicts (likely either me, you, or @BoxyUwU).

I personally think moving each project forward independently and merging them when ready is the right call. Merge conflicts aren't fun, but I don't think they should be a show stopper here.

**Comment by Frizi** [comment]
I've added very basic implementation for static storages. Now derive understands `#[storage(sparse)]` attribute (the syntax will probably change). That static storage data is used to transform all `fn is_dense` into associated `const DENSE: bool`. That way its evaluation is forced to be compile-time. That means iterators don't even store `is_dense` field, as the const is used directly instead. This allows the unused code path to be trivially eliminated.

Some benchmark numbers:
```
group                                     runtime is_dense (aafa863)  const IS_DENSE (40711df)
-----                                     ----------                 ------------
add_remove_component/bevy_sparse_set      1.04  1407.9±196.77µs      1.00  1348.9±84.43µs
add_remove_component/bevy_table           1.00  1699.0±57.21µs       1.00  1699.8±167.69µs
add_remove_component_big/bevy_sparse_set  1.00  1545.9±281.94µs      1.04  1611.1±193.76µs
add_remove_component_big/bevy_table       1.02      4.0±0.16ms       1.00      3.9±0.09ms
fragmented_iter/bevy                      1.14    450.5±9.97ns       1.00   394.6±31.22ns
fragmented_iter/bevy_foreach              1.00   310.0±20.17ns       1.02   315.9±24.39ns
get_component/bevy                        1.00  1495.5±106.58µs      1.00  1499.3±208.11µs
get_component/bevy_system                 1.00  1194.1±48.76µs       1.01  1211.0±85.75µs
heavy_compute/bevy                        1.06  1140.0±174.05µs      1.00  1070.5±22.27µs
schedule/bevy                             1.02     60.0±3.09µs       1.00     58.8±2.87µs
simple_insert/bevy                        1.03   647.8±56.01µs       1.00   627.0±10.97µs
simple_insert/bevy_unbatched              1.05  1602.2±160.03µs      1.00  1529.1±52.90µs
simple_iter/bevy                          1.28     21.5±0.74µs       1.00     16.9±0.77µs
simple_iter/bevy_foreach                  1.00     12.5±0.32µs       1.01     12.6±0.58µs
simple_iter/bevy_sparse                   1.00     63.4±7.15µs       1.00     63.6±3.73µs
simple_iter/bevy_sparse_foreach           1.18    61.2±10.73µs       1.00     51.8±2.07µs
sparse_fragmented_iter/bevy               1.13     15.1±0.28ns       1.00     13.4±0.48ns
sparse_fragmented_iter/bevy_foreach       1.13     12.8±0.42ns       1.00     11.3±1.42ns
```

This implementation is actually fairly straightforward and doesn't introduce any hard to resolve conflicts. In the future, we could do more in-depth refactoring to simplify the iterator implementations and deduplicate some code.

**Review by TheRawMeatball — APPROVED** [review]

File: crates/bevy_ecs/macros/src/lib.rs
Comment (TheRawMeatball) [inline crates/bevy_ecs/macros/src/lib.rs:160]: I'm unsure how to update this, can someone else please take a look?
cc @bjorn3 , it seems you removed TypeInfo

Comment (bjorn3) [reply crates/bevy_ecs/macros/src/lib.rs:160]: `ComponentId` is enough, right?

File: crates/bevy_ecs/src/component.rs
Comment (cart) [inline crates/bevy_ecs/src/component.rs:184]: We should probably add a SAFETY line above the `storage_type` field in the struct definition (like we do for `is_send_and_sync`). We can't let people override this / set it to something mismatched.

File: crates/bevy_ecs/src/lib.rs
Comment (cart) [inline crates/bevy_ecs/src/lib.rs:1342]: This is no longer a sparse drop check / it is now redundant with the previous test. We should use a DropCkSparse component here.

File: crates/bevy_ecs/README.md
Comment (cart) [inline crates/bevy_ecs/README.md:196]: The component_storage example no longer exists.

File: crates/bevy_sprite/src/lib.rs
Comment (cart) [inline crates/bevy_sprite/src/lib.rs:91]: We should flatten these lines into `let world_cell = app.world.cell()`.

File: crates/bevy_ecs/macros/src/lib.rs
Comment (cart) [reply crates/bevy_ecs/macros/src/lib.rs:160]: In this case, safety hinges on component_ids being returned in "field-definition-order" (and they are). Just swap TypeInfo in the `// SAFE:` section for ComponentIds and we're good to go.

File: crates/bevy_ecs/src/bundle.rs
Comment (cart) [inline crates/bevy_ecs/src/bundle.rs:79]: This doesn't appear to be used anywhere. Can we remove it?

File: crates/bevy_ecs/src/query/fetch.rs
Comment (cart) [inline crates/bevy_ecs/src/query/fetch.rs:236]: Adding a registration here (and not elsewhere, like WriteState) seems a bit odd. Maybe we need some sort of `world.init_component::<T>() -> ComponentId` function that does a get_or_insert operation _and_ sets up storages if necessary?  Maybe nothing should call `components.get_or_insert_info::<T>()` directly?

**Review by cart — COMMENTED** [review]
This pretty much looks ready to me. Just some small comments here and there.

File: crates/bevy_ecs/src/world/mod.rs
Comment (cart) [inline crates/bevy_ecs/src/world/mod.rs:194]: I dont think we should be constructing a ComponentDescriptor (which allocates a String) every time something needs a ComponentId.

Ideally we only initialize the descriptor if the component hasn't already been registered.

**Comment by cart** [comment]
Just pushed a change that centralizes and simplifies component initialization (and resolves my comment here: https://github.com/bevyengine/bevy/pull/2254#discussion_r720562185). Its relatively straightforward, but I'd appreciate it if at least one other person reviewed it.

After that I think this is good to merge!

**Review by TheRawMeatball — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]

**Pull request closed without merging**
